src/avl.h

```diff
@@ -22,64 +42,122 @@
- * a is linked directly to the tree, so it has to
- * be properly allocated by the caller.
+// ----------------------------------------------------------------------------
+// Public methods
+
+/** 
+ * Insert `a` into AVL tree `t`.
+ *
+ * Insert element `a` into the AVL tree `t`.
+ * If an element equal (as returned by `t->compar()`) to `a` exits in the tree, `a` is not added.
+ * `a` is linked directly to the tree, so it has to be properly allocated by the caller.
+ *
+ * @param t avl tree
+ * @param a element to insert
+ * @retrun `a` or an equal element
  */
 avl *avl_insert_lock(avl_tree_lock *t, avl *a) NEVERNULL WARNUNUSED;
+/** 
+ * Insert `a` into AVL tree `t`.
+ *
+ * Insert element `a` into the AVL tree `t`.
+ * If an element equal (as returned by `t->compar()`) to `a` exits in the tree, `a` is not added.
+ * `a` is linked directly to the tree, so it has to be properly allocated by the caller.
+ *
+ * @param t AVL tree
+ * @param a element to insert
+ * @retrun `a` or an equal element
+ */
 avl *avl_insert(avl_tree *t, avl *a) NEVERNULL WARNUNUSED;
 
-/* Remove an element a from the AVL tree t
- * returns a pointer to the removed element
- * or NULL if an element equal to a is not found
- * (equal as returned by t->compar())
+/** 
+ * Remove `a` from AVL tree `t`.
+ *
+ * Remove an element equal (as returned by `t->compar()`) to `a` from the AVL tree `t`.
+ *
+ * @param t AVL tree
+ * @param a element to remove
+ * @return the removed element or NULL if not found
  */
 avl *avl_remove_lock(avl_tree_lock *t, avl *a) WARNUNUSED;
+/** 
+ * Remove `a` from AVL tree `t`.
+ *
+ * Remove an element equal (as returned by `t->compar()`) to `a` from the AVL tree `t`.
+ *
+ * @param t AVL tree
+ * @param a element to remove
+ * @return the removed element or NULL if not found
+ */
 avl *avl_remove(avl_tree *t, avl *a) WARNUNUSED;
 
-/* Find the element into the tree that equal to a
- * (equal as returned by t->compar())
- * returns NULL is no element is equal to a
+/** 
+ * Find `a` in AVL tree `t`.
+ * 
+ * Find a element equal (as returned by `t->compar()`) to `a` in AVL tree `t`.
+ *
+ * @param a element to find
+ * @param t AVL tree
+ * @return found element or NULL
  */
 avl *avl_search_lock(avl_tree_lock *t, avl *a);
+/** 
+ * Find `a` in AVL tree `t`.
+ * 
+ * Find a element equal (as returned by `t->compar()`) to `a` in AVL tree `t`.
+ *
+ * @param a element to find
+ * @param t AVL tree
+ * @return found element or NULL
+ */
 avl *avl_search(avl_tree *t, avl *a);
 
-/* Initialize the avl_tree_lock
+/**
+ * Initialize a avl_tree_lock.
+ *
+ * @param t AVL tree
+ * @param compar() Comparison function
  */
 void avl_init_lock(avl_tree_lock *t, int (*compar)(void *a, void *b));
+/**
+ * Initialize a avl_tree_lock.
+ *
+ * @param t AVL tree
+ * @param compar() Comparison function
+ */
 void avl_init(avl_tree *t, int (*compar)(void *a, void *b));
 
-
+/**
+ * \todo kstou Your help needed.
```

Comment:
walk through all entries in the AVL tree. The function `int callback(entry, data)` will be called for each node (`data` is the `data` parameter to the traverse function).

The return value of `callback()` is interpreted as follows:

- zero, or positive = add it to `total`.
- negative = stop traversing the tree.

If no call to callback returns a negative number, the return value of the traverse function is the `total`. 
If any call to callback returns a negative number, the return value is the number returned by that callback function.

I have never verified, if this calls the callback function in a sorted order (I hope it does, but I never really cared since netdata uses these AVL trees with string hashes - so sorting them based on the hashes is meaningless).